flox/core.py

```diff
@@ -468,8 +468,26 @@ def factorize_(
                 bins = np.concatenate([expect.left.to_numpy(), [expect.right[-1].to_numpy()]])
             else:
                 bins = np.concatenate([expect.left.to_numpy(), [expect.right[-1]]])
+
             # code is -1 for values outside the bounds of all intervals
-            idx = pd.cut(flat, bins=bins, right=expect.closed_right).codes.copy()
+            # idx = pd.cut(flat, bins=bins, right=expect.closed_right).codes.copy()
+
+            # digitize is 0 or idx.max() for values outside the bounds of all intervals
+            # make it behave like pd.cut:
+            if len(bins) > 1:
+                right = expect.closed_right
+                idx = np.digitize(
+                    flat,
+                    bins=bins.view(np.intp) if bins.dtype.kind == "M" else bins,
+                    right=right,
+                )
+                idx = pd.to_numeric(idx, downcast="integer")
+                idx -= 1
+                within_bins = flat <= bins.max() if right else flat < bins.max()
+                idx[~within_bins] = -1
+            else:
```

Comment:
Feel like this `else` should just be an error? Would `pd.cut` succeed?

Reply:
Yes, both succeeds in their own way and I mostly just aimed for exact replacement of pd.cut:
```python
data = [2, 4, 4, 7, 12, 14, 20, 22, 24, 31, 34]

np.digitize(data, bins=[10, 20, 30], right=True)
Out[15]: array([0, 0, 0, 0, 1, 1, 1, 2, 2, 3, 3], dtype=int64)
pd.cut(data, bins=[10, 20, 30], right=True).codes
Out[16]: array([-1, -1, -1, -1,  0,  0,  0,  1,  1, -1, -1], dtype=int8)

np.digitize(data, bins=[10, 20, 40], right=True)
Out[13]: array([0, 0, 0, 0, 1, 1, 1, 2, 2, 2, 2], dtype=int64)
pd.cut(data, bins=[10, 20, 40], right=True).codes
Out[14]: array([-1, -1, -1, -1,  0,  0,  0,  1,  1,  1,  1], dtype=int8)

np.digitize(data, bins=[10, 20], right=True)
Out[7]: array([0, 0, 0, 0, 1, 1, 1, 2, 2, 2, 2], dtype=int64)
pd.cut(data, bins=[10, 20], right=True).codes
Out[8]: array([-1, -1, -1, -1,  0,  0,  0, -1, -1, -1, -1], dtype=int8)

np.digitize(data, bins=[10], right=True)
Out[9]: array([0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1], dtype=int64)
pd.cut(data, bins=[10], right=True).codes
Out[10]: array([-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1], dtype=int8)

np.digitize(data, bins=[], right=True)
Out[11]: array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int64)
pd.cut(data, bins=[], right=True).codes
Out[12]: array([-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1], dtype=int8)
```